session/session_test.go

```diff
@@ -5149,3 +5150,232 @@ func (s *testSessionSuite) TestLocalTemporaryTableScan(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("set @@tidb_enable_noop_functions=1")
+	tk.MustExec("use test")
+
+	tk.MustExec("create temporary table tmp1 (id int primary key, u int unique, v int)")
+
+	insertRecords := func(idList []int) {
+		for _, id := range idList {
+			tk.MustExec("insert into tmp1 values (?, ?, ?)", id, id+100, id+1000)
+		}
+	}
+
+	idList := []int{1, 2, 3, 4, 5, 6, 7, 8, 9}
+
+	checkUpdatesAndDeletes := func(updates []string, deletes []int) {
+		modifyMap := make(map[int]string)
+		for _, m := range updates {
+			parts := strings.Split(strings.TrimSpace(m), " ")
+			c.Assert(len(parts) != 0, IsTrue)
+			id, err := strconv.Atoi(parts[0])
+			c.Assert(err, IsNil)
+			modifyMap[id] = m
+		}
+
+		for _, d := range deletes {
+			modifyMap[d] = ""
+		}
+
+		expect := make([]string, 0)
+		for _, id := range idList {
+			modify, exist := modifyMap[id]
+			if !exist {
+				expect = append(expect, fmt.Sprintf("%d %d %d", id, id+100, id+1000))
+				continue
+			}
+
+			if modify != "" {
+				expect = append(expect, modify)
+			}
+
+			delete(modifyMap, id)
+		}
+
+		otherIds := make([]int, 0)
+		for id := range modifyMap {
+			otherIds = append(otherIds, id)
+		}
+
+		sort.Ints(otherIds)
+		for _, id := range otherIds {
+			modify, exist := modifyMap[id]
+			c.Assert(exist, IsTrue)
+			expect = append(expect, modify)
+		}
+
+		tk.MustQuery("select * from tmp1").Check(testkit.Rows(expect...))
+	}
+
+	executeUpdateSql := func(sql string, expectedUpdateSuccess bool) (err error) {
+		if expectedUpdateSuccess {
+			tk.MustExec(sql)
+		} else {
+			err = tk.ExecToErr(sql)
+			c.Assert(err, NotNil)
+			checkUpdatesAndDeletes(nil, nil)
+		}
+		tk.MustQuery("show warnings").Check(testkit.Rows())
+		return
+	}
+
+	assertUpdate := func(sql string, expectedUpdateSuccess bool, verify func(err error)) {
+		// update records in txn and records are inserted in txn
+		tk.MustExec("begin")
+		insertRecords(idList)
+		err := executeUpdateSql(sql, expectedUpdateSuccess)
+		verify(err)
+		tk.MustExec("rollback")
+		tk.MustQuery("select * from tmp1").Check(testkit.Rows())
+
+		// update records out of txn
+		insertRecords(idList)
+		err = executeUpdateSql(sql, expectedUpdateSuccess)
+		verify(err)
+		tk.MustExec("delete from tmp1")
+
+		// update records in txn and rollback
+		insertRecords(idList)
+		tk.MustExec("begin")
+		err = executeUpdateSql(sql, expectedUpdateSuccess)
+		verify(err)
+		tk.MustExec("rollback")
+		// rollback left records unmodified
+		checkUpdatesAndDeletes(nil, nil)
+
+		// update records in txn and commit
+		tk.MustExec("begin")
+		err = executeUpdateSql(sql, expectedUpdateSuccess)
+		verify(err)
+		tk.MustExec("commit")
+		if !expectedUpdateSuccess {
+			checkUpdatesAndDeletes(nil, nil)
+		}
+		verify(err)
+		tk.MustExec("delete from tmp1")
+		tk.MustQuery("select * from tmp1").Check(testkit.Rows())
+	}
+
+	// update with point get for primary key
+	assertUpdate("update tmp1 set v=999 where id=1", true, func(_ error) {
```

Comment:
Can you make the tests table driven, and use less callback function as possible?

For example, 

```
type TestCase struct {
    sql string
    expected string
}

for _, c := range testCases {
    testCase(c)
}

func testCase() {
    // the assertUpdate function
}
```
`checkUpdatesAndDeletes` do not really need to be a callback function, most of the case do not need that flexible... If you really want that, you can make it data oriented, for example, 

```
 testCases := []testCase{
    {"update tmp1 set id=12 where id=1",  nil},
    {"update tmp1 set id=12 where id=1", checkResult{12, 101, 1001}},
    {"update tmp1 set id=1 where id=1", checkResult{}},
    {},
}


type checkResult  struct {
   ...
}

func (checkResult) Check() {
  .....
}
```

In this way, the test cases are driven by a table, and if you think the pure data is not flexible enough, you can make them into data struct and wrap method on them (as the callback function) ...